vdr/didx509/resolver.go

```diff
@@ -0,0 +1,232 @@
+	if err != nil {
+		return nil, nil, err
+	}
+	validationCert, err := findCertificateByHash(chain, metadata.X509CertThumbprint, "sha1")
+	if err != nil {
+		validationCert, err = findCertificateByHash(chain, metadata.X509CertThumbprintS256, "sha256")
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	err = validatePolicy(ref, validationCert)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	err = r.pkiValidator.Validate(chain)
+	if err != nil {
+		return nil, nil, err
+	}
+	document, err := createDidDocument(id, validationCert)
+	if err != nil {
+		return nil, nil, err
+	}
+	return document, &resolver.DocumentMetadata{}, err
+}
+
+// validatePolicy validates a certificate against a given X509DidReference and its policy.
+func validatePolicy(ref *X509DidReference, cert *x509.Certificate) error {
+
+	switch ref.PolicyName {
+	case "subject":
+		keyValue := strings.Split(ref.PolicyValue, ":")
+		if len(keyValue)%2 != 0 {
+			return ErrDidSanMalformed
+		}
+		for i := 0; i < len(keyValue); i = i + 2 {
+			subject := cert.Subject
+			key := keyValue[i]
+			value := keyValue[i+1]
+			switch key {
+			case "serialNumber":
+				if subject.SerialNumber != value {
+					return ErrWrongSerialNumber
+				}
+			case "CN":
+				if subject.CommonName != value {
+					return ErrWrongCN
+				}
+			case "L":
+				if !slices.Contains(subject.Locality, value) {
+					return ErrWrongLocality
+				}
+			case "O":
+				if !slices.Contains(subject.Organization, value) {
+					return ErrWrongOrganization
+				}
+			case "OU":
+				if !slices.Contains(subject.OrganizationalUnit, value) {
+					return ErrWrongOrganizationalUnit
+				}
+			}
+		}
+	case "san":
+		keyValue := strings.Split(ref.PolicyValue, ":")
+		if len(keyValue)%2 != 0 {
+			return ErrDidSanMalformed
+		}
+		for i := 0; i < len(keyValue); i = i + 2 {
+			key := keyValue[i]
+			value := keyValue[i+1]
+			switch key {
+			case "otherName":
+				nameValue, err := findOtherNameValue(cert)
+				if err != nil {
+					return err
+				}
+				if nameValue != value {
+					return ErrWrongSanOtherName
+				}
+			case "dns":
+				if len(cert.DNSNames) > 0 && !slices.Contains(cert.DNSNames, value) {
+					return ErrWrongSanDns
+				}
+			case "email":
+				if len(cert.EmailAddresses) > 0 && !slices.Contains(cert.EmailAddresses, value) {
+					return ErrWrongSanEmailAddresses
+				}
+			case "ip":
+				if len(cert.IPAddresses) > 0 {
+					ok := false
+					for _, ip := range cert.IPAddresses {
+						if ip.String() == value {
+							ok = true
+							break
+						}
+					}
+					if !ok {
+						return ErrWrongSanIPAddresses
+					}
+
+				}
+			}
+
+		}
+	}
+
+	return nil
+}
+
+// createDidDocument generates a new DID Document based on the provided DID identifier and validation certificate.
+func createDidDocument(id did.DID, validationCert *x509.Certificate) (*did.Document, error) {
+	verificationMethod, err := did.NewVerificationMethod(did.DIDURL{DID: id}, ssi.JsonWebKey2020, id, validationCert.PublicKey)
```

Comment:
The ID of the verification method needs to be the DID (you ahve this), appended with the `#0` fragment. I think you can set `Fragment` on the first argument to `0` to achieve this.

Reply:
Set the fragment, added a check to the tests.